articles/kaneshin.re

```diff
@@ -0,0 +1,413 @@
+  logger.a.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+
+func main() {
+  runtime.Callers(1, pcs[:])
+  logger.main.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+//}
+
+//list[initialize_package_a][a.go][go]{
+package main
+
+import "runtime"
+
+var _ = suba()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.a.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func suba() int {
+  runtime.Callers(1, pcs[:])
+  logger.a.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+//}
+
+//list[initialize_package_z][z.go][go]{
+package main
+
+import "runtime"
+
+var _ = subz()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.z.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func subz() int {
+  runtime.Callers(1, pcs[:])
+  logger.z.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+//}
+
+ファイル名は、ファイル名の昇順でわかりやすいものを定義しているとの、出力と呼び出された関数オブジェクトの名前を取得するために下記の変数を定義しています。
+
+//list[initialize_package_logger][][go]{
+var logger = struct {
+  a    *log.Logger
+  z    *log.Logger
+  main *log.Logger
+}{
+  a:    log.New(os.Stdout, "   a.go: ", 0),
+  z:    log.New(os.Stdout, "   z.go: ", 0),
+  main: log.New(os.Stdout, "main.go: ", 0),
+}
+
+var pcs [1]uintptr
+//}
+
+//cmd{
+$ go build -o main . && ./main
+   a.go: main.suba
+   a.go: main.submain
+   z.go: main.subz
+   a.go: main.init.0
+main.go: main.init.1
+   z.go: main.init.2
+main.go: main.main
+//}
+
+それぞれのファイルにある変数の初期化が処理され、その後に@<code>{init}関数が処理されています。@<code>{main}関数のあるファイルが最後に初期化処理されるわけではありませんし、main.goやinit.goというファイル名を作ってもファイル名による優先処理はありません。
+
+==== 依存パッケージの初期化
+
+あるパッケージに別のパッケージが依存している場合、依存パッケージが先に初期化処理をするように実行されます。この初期化は再帰的に行われるため、依存の依存パッケージが存在すれば先に依存の依存パッケージの初期化処理がされます。
+
+//list[initialize_package_dep_a][dep/a.go][go]{
+package dep
+
+import "log"
+import "os"
+import "runtime"
+
+var logger = log.New(os.Stdout, " d/a.go: ", 0)
+
+var pcs [1]uintptr
+
+var _ = suba()
+
+func init() {
+  runtime.Callers(1, pcs[:])
+  logger.Println(runtime.FuncForPC(pcs[0]).Name())
+}
+
+func suba() int {
+  runtime.Callers(1, pcs[:])
+  logger.Println(runtime.FuncForPC(pcs[0]).Name())
+  return 0
+}
+//}
+
+@<code>{import _ "github.com/kaneshin/foo/bar/dep"}のようにブランクインポートさせると下記のように先に依存パッケージの初期化処理が行われて、依存元のパッケージの初期化が処理されます。
+
+//cmd{
+$ go build -o main . && ./main
+ d/a.go: github.com/kaneshin/foo/bar/dep.suba
+ d/a.go: github.com/kaneshin/foo/bar/dep.init.0
+   a.go: main.suba
+   a.go: main.submain
+   z.go: main.subz
+   a.go: main.init.0
+main.go: main.init.1
+   z.go: main.init.2
+main.go: main.main
+//}
+
+==== 初期化のコントロール
+
+初期化を完全に管理下に置きたい場合はinit.goなどにまとめるのが良いですが、変数や@<code>{init}関数、そして依存パッケージも上から順に再帰的に処理されます。このように順序関係を意識する必要のある初期化処理や、他の初期化処理に依存する初期化処理が必要となる設計はそもそも見直しが必要です。すべてを実行時に初期化処理することができれば良いかもしれませんが、ある程度妥協することも設計の上では大事になります。複雑性を解決するためにGoが開発されたのに、逆に複雑にしては元も子もありません。
+
+== パッケージスコープの変数宣言と初期化
+
+パッケージスコープの変数を宣言させることはよいことなのか？Goは並行処理される言語のためパッケージスコープの変数へは競合状態を気にする必要があります。パッケージスコープな変数は@<code>{goroutine}間での処理を意識しながら開発と実行をしなくてはならなくなるため、密結合の温床であり依存性の高いコードとなるのであまり推奨されません。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "？". (ja-technical-writing/no-exclamation-question-mark)